### ISSUE TITLE
ai::NeuralNetwork backpropagates through weights it has already updated

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#130]
`NeuralNetwork<T>::train` (`jlib/ai/neural.hh`) applies each layer's weight
update and *then* reads that layer's weights to backpropagate through:

```cpp
m_who += rate * (((output_errors ^ final_outputs ^ (1.0 - final_outputs))
                  * deep_outputs.transpose()));

math::matrix<T> deep_errors = output_errors;
math::matrix<T> deep = m_who;              // <-- after the update

for(int i = m_deep.size() - 1; i >= 0; i--) {
    deep_errors = deep.transpose() * deep_errors;
    m_deep[i] += rate * (...);
    deep = m_deep[i];                      // <-- likewise
}

math::matrix<T> hidden_errors = deep.transpose() * deep_errors;
```

So the error arriving at the hidden layer has been propagated through weights
that already moved. Backpropagation evaluates every gradient at the same
weights and applies them afterwards; this evaluates the later ones at weights
partway through the step.

## Measured

From identical starting weights, with two samples: train net A on sample 0
alone, net B on sample 1 alone, and net C on both as a batch of two. Since a
batch gradient is the mean, C's step should be the average of A's and B's, and
all three gradients are taken at the same weights, so it should be exact.

```
  who (output layer): worst |batch - mean| = 5.551e-17
  wih (hidden layer): worst |batch - mean| = 1.535e-06
```

The output layer is exact to double rounding. The hidden layer is out by eleven
orders of magnitude more, and that gap is the already-updated `m_who` in the
backward pass.

## Why this is recorded rather than fixed

The fix is small -- capture each layer's weights before updating it, or apply
all the updates after the backward pass -- but it changes the arithmetic of
every training step, so every existing network trains differently afterwards.
There is no convergence test in the tree to say whether that is better, worse
or indistinguishable, and `jneural-alpha` needs MNIST to answer it.

It has presumably always been this way, and networks do train, so it is a
deviation rather than a disaster. It matters more now than it did: it is the
reason a batch of N is not exactly the mean of N single steps, which is the
property batching is supposed to have.

## What a fix would need

- Capture `m_who` and each `m_deep[i]` before updating them.
- A convergence check with a real dataset, before and after, since the whole
  claim is about training behaviour rather than about a shape.
- A decision about whether to keep the current behaviour available, given
  anyone with saved weights trained under it.

`tests/ai_neural_test.cc` asserts the split as it stands -- the output layer
exactly, the hidden layer approximately -- so that whoever fixes this can see
which half was already right.

Found while adding batch support (mean gradient), which is what made the
inexactness visible.

**Comment by xoloki** [comment]
## How wrong, quantified

Worth putting a number on it before anyone decides this is urgent.

```
  lrate 0.01   largest wih update  1.454e-04
  lrate 0.10   largest wih update  1.449e-03
  lrate 0.50   largest wih update  7.129e-03
  lrate 1.00   largest wih update  1.397e-02
```

The batch-versus-mean gap measured at lrate 0.1 was **1.535e-06**, against a
largest update of 1.449e-03 -- so the deviation is about **0.1% of one step**.

That is the shape to expect: `m_who` moves by O(lrate * gradient) before being
used in the backward pass, so the hidden gradient is off by O(lrate) and the
error vanishes as the learning rate shrinks. It is a perturbed descent
direction, not a wrong one, which is why networks trained under it anyway --
`jneural-alpha` learned MNIST with this code.

Two places it should be expected to matter more, neither measured yet:

- **Larger learning rates**, since the perturbation scales linearly with the
  step size.
- **Depth.** The `m_deep` loop repeats the pattern at every layer -- `deep =
  m_deep[i]` after `m_deep[i] +=` -- so the staleness compounds one hop per
  layer rather than happening once. One hidden layer is a single 0.1% nudge;
  `jneural-search` builds arbitrary depth and would accumulate it.

So the priority reads as low for the shallow networks in the tree and unknown
for deep ones, and the deep case is the one worth measuring before this is
either fixed or closed as immaterial.

**Comment by xoloki** [comment]
## I was wrong about the severity, and by a lot

The comment above put this at "about 0.1% of one step" and called it a
perturbed descent direction rather than a wrong one. That figure was measured on
a network with **one hidden layer** -- the case where an error that compounds
once per layer compounds least. Nobody measured it at depth, including me.

Measured now, on MNIST with relu hidden layers, 64 wide, two epochs:

```
                    unfixed                    fixed
  rate:     0.003   0.001  0.0003     0.003   0.001  0.0003
  3 x 64     9.8%    9.8%  88.08%    96.06%  94.29%  88.34%
  4 x 64     9.8%    9.8%   9.80%     9.80%   9.80%  70.21%
  6 x 64     9.8%    9.8%   9.80%     9.80%   9.80%   9.80%
```

**Three hidden layers: 88.08% to 96.06%**, and it now trains at a learning rate
ten times larger -- 0.003, which was previously chance. **Four hidden layers:
chance to 70.21%.** The wall that #131's activation work could not break moves
one layer out as soon as this is fixed.

So this is not a rounding-order curiosity. It is a first-order error that
compounds once per layer, and past two or three layers it dominates.

## The fix that produced those numbers

Capture each layer's weights before updating them, so every gradient is taken at
the weights the step started from:

```cpp
math::matrix<T> deep = m_who;          // before, not after

m_who += rate * (...);

for(int i = m_deep.size() - 1; i >= 0; i--) {
    deep_errors = deep.transpose() * deep_errors;

    math::matrix<T> before = m_deep[i];

    m_deep[i] += rate * (...);
    deep = before;
}
```

One extra matrix copy per layer per step, which against the multiplies around it
is nothing.

## Caveats on those numbers

Two epochs, one seed, one width, and `jneural-alpha` is default-seeded so each
cell is a single deterministic run rather than an average. The direction and
the magnitude are not in doubt -- chance to 70% is not a sampling artefact --
but the exact figures should not be quoted as benchmarks.

Also unchanged: six layers still sits at chance at every rate tried. Whatever
stops it is a third thing, after this and the activation.

## Why the first measurement misled

The single-layer measurement compared a batch step against the mean of two
single steps and found them 1.5e-6 apart. That is a true statement about one
layer and a useless predictor of four, because the quantity it measures is not
the one that compounds. A better early check would have been to ask what the
error does as depth grows, rather than how big it is at the depth that was
convenient to measure.

Thanks to @xoloki for pushing back on the dismissal.
